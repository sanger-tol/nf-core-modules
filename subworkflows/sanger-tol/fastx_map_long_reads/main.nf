include { FASTXALIGN_PYFASTXINDEX    } from '../../../modules/sanger-tol/fastxalign/pyfastxindex/main'
include { FASTXALIGN_MINIMAP2ALIGN   } from '../../../modules/sanger-tol/fastxalign/minimap2align/main'
include { MINIMAP2_INDEX             } from '../../../modules/nf-core/minimap2/index/main'

include { BAM_SAMTOOLS_MERGE_MARKDUP } from '../bam_samtools_merge_markdup/main'

workflow FASTX_MAP_LONG_READS {

    take:
    ch_assemblies             // Channel [meta, assembly]
    ch_fasta                  // Channel [meta, fasta] OR [meta, [fasta1, fasta2, ..., fasta_n]]
    val_reads_per_fasta_chunk // integer: Number of reads per FASTA chunk for mapping
    val_output_bam            // boolean: if true output alignments in BAM format

    main:
    ch_versions = Channel.empty()

    //
    // Module: Index FASTA files
    //
    FASTXALIGN_PYFASTXINDEX(ch_fasta.transpose())
    ch_versions = ch_versions.mix(FASTXALIGN_PYFASTXINDEX.out.versions)

    //
    // Logic: Identify FASTA chunks
    //
    ch_fastx_chunks = FASTXALIGN_PYFASTXINDEX.out.index
        | map { meta, index, count ->
            def intcount = count.toInteger()
            def size     = val_reads_per_fasta_chunk
            def n_bins   = Math.ceil(intcount / size).toInteger()
            chunkn       = (0..<n_bins).collect()
            slices       = chunkn.collect { chunk ->
                def lower = chunk * size
                def upper = [lower + size, intcount].min()

                return [ lower, upper ]
            }

            return [ meta, index, chunkn, slices ]
        }

    //
    // Logic: Count the total number of cram chunks for downstream grouping
    //
    ch_n_fasta_chunks = ch_fastx_chunks
        | map { meta, index, chunkn, _slices -> [ meta, chunkn ] }
        | transpose()
        | groupTuple(by: 0)
        | map { meta, chunkns -> [ meta, chunkns.size() ] }

    //
    // MODULE: generate minimap2 mmi file
    //
    MINIMAP2_INDEX(ch_assemblies)
    ch_versions = ch_versions.mix(MINIMAP2_INDEX.out.versions)

    //
    // Module: Map slices of each FASTA file to the reference
    //
    ch_fasta_with_slices = ch_fasta
        | combine(ch_fastx_chunks, by: 0)
        | combine(MINIMAP2_INDEX.out.index, by: 0)
        | transpose()

    FASTXALIGN_MINIMAP2ALIGN(
        ch_fasta_with_slices,
        val_output_bam
    )
    ch_versions = ch_versions.mix(FASTXALIGN_MINIMAP2ALIGN.out.versions)

    //
    // Logic: Group all PAF files together, using a groupKey to output when
    //        reaching the expected count of PAF files
    //
    ch_grouped_paf = FASTXALIGN_MINIMAP2ALIGN.out.paf
        | combine(ch_n_fasta_chunks, by: 0)
        | map { meta, paf, n_chunks ->
            def key = groupKey(meta, n_chunks)
            [key, paf]
        }
        | groupTuple(by: 0, sort: { it.getName() } )
        | map { key, paf -> [key.target, paf] } // Get meta back out of groupKey

    //
    // Logic: Group all BAM files together for merging, using a groupKey to
    //        output when reaching the expected count of PAF files
    //
    ch_merge_input = FASTXALIGN_MINIMAP2ALIGN.out.bam
        | combine(ch_n_fasta_chunks, by: 0)
        | map { meta, bam, n_chunks ->
            def key = groupKey(meta, n_chunks)
            [key, bam]
        }
        | groupTuple(by: 0)
        | map { key, bam -> [key.target, bam] } // Get meta back out of groupKey

    //
    // Logic: Wrap this in the conditional so we don't unnecessarily run
    //        samtools faidx if no bam output
    //
    ch_output_bam = Channel.empty()

    if(val_output_bam) {
        BAM_SAMTOOLS_MERGE_MARKDUP(
            ch_merge_input,
            ch_assemblies,
            false
        )
        ch_versions = ch_versions.mix(BAM_SAMTOOLS_MERGE_MARKDUP.out.versions)

        ch_output_bam = ch_output_bam.mix(BAM_SAMTOOLS_MERGE_MARKDUP.out.bam)
    }

    emit:
    bam      = ch_output_bam
    paf      = ch_grouped_paf
    versions = ch_versions
}
