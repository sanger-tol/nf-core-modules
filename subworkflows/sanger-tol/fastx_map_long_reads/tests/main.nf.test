nextflow_workflow {

    name "Test Subworkflow FASTX_MAP_LONG_READS"
    script "../main.nf"
    workflow "FASTX_MAP_LONG_READS"

    tag "subworkflows"
    tag "subworkflows_sangertol"
    tag "subworkflows/fastx_map_long_reads"
    tag "fastxalign"
    tag "fastxalign/pyfastxindex"
    tag "fastxalign/minimap2align"
    tag "subworkflows/../../modules/nf-core/minimap2/index"
    tag "subworkflows/bam_samtools_merge_markdup"


    config "./nextflow.config"

    setup {
        nfcoreInitialise("${launchDir}/library/")
        nfcoreInstall(
            "${launchDir}/library/",
            [
                "minimap2/index",
                "samtools/merge",
                "samtools/markdup",
                "samtools/faidx"
            ]
        )
        nfcoreLink("${launchDir}/library/", "${baseDir}/modules/")
    }

    test("Undibacterium unclassified - fasta - paf") {

        when {

            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ],
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/assembly/draft/baUndUnlc1.hic.hap1.p_ctg.unscaffolded.fa.gz', checkIfExists: true),
                ])
                input[1] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/genomic_data/baUndUnlc1/pacbio/fasta/HiFi.reads.fasta', checkIfExists: true),
                ])
                input[2] = 200
                input[3] = false
                """
            }
        }
        then {
            assert workflow.success
            assertAll(
                { assert snapshot(
                    workflow.out.paf,
                    workflow.out.versions
                ).match() }
            )
        }
        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }

    test("Undibacterium unclassified - fasta - bam") {

        when {

            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ],
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/assembly/draft/baUndUnlc1.hic.hap1.p_ctg.unscaffolded.fa.gz', checkIfExists: true),
                ])
                input[1] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/genomic_data/baUndUnlc1/pacbio/fasta/HiFi.reads.fasta', checkIfExists: true),
                ])
                input[2] = 200
                input[3] = true
                """
            }
        }
        then {
            assert workflow.success
            assertAll(
                { assert snapshot(
                    workflow.out.bam.collect { bam(it.get(1)).getStatistics() },
                    workflow.out.versions
                ).match() }
            )
        }
        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }

}
