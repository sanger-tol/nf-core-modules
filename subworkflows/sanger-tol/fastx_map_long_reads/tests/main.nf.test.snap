{
    "Undibacterium unclassified - fasta - bam": {
        "content": [
            [
                {
                    "maxReadLength": 26402,
                    "minReadLength": 0,
                    "meanReadLength": 7644,
                    "maxQuality": 60,
                    "minQuality": 0,
                    "meanQuality": 54,
                    "readCount": 1038,
                    "duplicateReadCount": 0,
                    "sorted": false
                }
            ],
            [
                "versions.yml:md5,0c140183a1240ea56d0d489501143554",
                "versions.yml:md5,0f8abad8666548b83cdf8a2087056aed",
                "versions.yml:md5,ab7cf2afd6c2bc60cfe7452fa4f5282b",
                "versions.yml:md5,e2202eedd003106e612336bcd8036b07",
                "versions.yml:md5,fb7f80773e78298912f0a6f3d80f0084",
                "versions.yml:md5,fb7f80773e78298912f0a6f3d80f0084",
                "versions.yml:md5,fb7f80773e78298912f0a6f3d80f0084",
                "versions.yml:md5,fb7f80773e78298912f0a6f3d80f0084",
                "versions.yml:md5,fb7f80773e78298912f0a6f3d80f0084"
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-22T09:12:34.001975"
    },
    "Undibacterium unclassified - fasta - paf": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "HiFi.reads.fasta.0.test.paf.gz:md5,d74ffe5ca051a58b834712bbfecb7875"
                ],
                [
                    {
                        "id": "test"
                    },
                    "HiFi.reads.fasta.1.test.paf.gz:md5,b534510e942d0b8ea7150bb08a0ac9e1"
                ],
                [
                    {
                        "id": "test"
                    },
                    "HiFi.reads.fasta.2.test.paf.gz:md5,ba98f31b846d1830279c51d3bc829760"
                ],
                [
                    {
                        "id": "test"
                    },
                    "HiFi.reads.fasta.3.test.paf.gz:md5,7f945c1697a0e727c672bffe0dab8799"
                ],
                [
                    {
                        "id": "test"
                    },
                    "HiFi.reads.fasta.4.test.paf.gz:md5,ee7fdba4b0311d309292f28ada0ef137"
                ]
            ],
            [
                "versions.yml:md5,0c140183a1240ea56d0d489501143554",
                "versions.yml:md5,ab7cf2afd6c2bc60cfe7452fa4f5282b",
                "versions.yml:md5,fb7f80773e78298912f0a6f3d80f0084",
                "versions.yml:md5,fb7f80773e78298912f0a6f3d80f0084",
                "versions.yml:md5,fb7f80773e78298912f0a6f3d80f0084",
                "versions.yml:md5,fb7f80773e78298912f0a6f3d80f0084",
                "versions.yml:md5,fb7f80773e78298912f0a6f3d80f0084"
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-22T09:11:33.158758"
    }
}